File: problems/leetcode/0100/124/124.cpp
// 124. Binary Tree Maximum Path Sum
// https://leetcode.com/problems/binary-tree-maximum-path-sum/description/

using namespace std;

#include <algorithm>
#include <climits>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  int maxSum = INT_MIN;

  int dfs(TreeNode *node) {
    int localMaxSum = node->val, left = INT_MIN, right = INT_MIN;

    if (node->left) {
      left = dfs(node->left);
      localMaxSum = max(localMaxSum, node->val + left);
    }
    if (node->right) {
      right = dfs(node->right);
      localMaxSum = max(localMaxSum, node->val + right);
    }

    int possiblePathSum = localMaxSum;
    if (left != INT_MIN && right != INT_MIN) {
      possiblePathSum = max(possiblePathSum, left + right + node->val);
    }
    maxSum = max(maxSum, possiblePathSum);

    return localMaxSum;
  }

  int maxPathSum(TreeNode *root) {
    dfs(root);
    return maxSum;
  }
};

// 한 쪽으로만 내려가면서 합의 최댓값을 안다면,
// 양 쪽의 합은 그 노드를 중심으로 하는 값의 최대값.
// 자기 자신, 양 쪽에서 한 쪽으로만 가는 것들의 최대값, 양 쪽 둘다 한 쪽으로만
// 가는 것들의 최대값의 합, 양 쪽에서 각각 알아서 하는 것들의 최대값

File: problems/leetcode/0100/124/124.py
# 124. Binary Tree Maximum Path Sum
# https://leetcode.com/problems/binary-tree-maximum-path-sum/description/

# (재사용 가능한 최대 합, 재사용 불가능한 최대 합)을 반환하는 함수가 있다고 가정하면

from typing import *

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        return max(self.reusableAndNonReusableMaxPathSum(root))

    def reusableAndNonReusableMaxPathSum(self, node: Optional[TreeNode]) -> tuple[float, float]:
        if not node:
            return 0, float("-inf")

        reusableLeft, nonReusableLeft = self.reusableAndNonReusableMaxPathSum(node.left)
        reusableRight, nonReusableRight = self.reusableAndNonReusableMaxPathSum(node.right)

        reusable = max(
            node.val, node.val + reusableLeft,
            node.val + reusableRight
        )
        nonReusable = max(
            reusableLeft if node.left else float("-inf"),
            reusableRight if node.right else float("-inf"),
            node.val + reusableLeft + reusableRight if node.left and node.right else float("-inf"),
            nonReusableLeft,
            nonReusableRight
        )

        return reusable, nonReusable
